MLP for NLP
------

Train and evaluate a simple MLP on the Reuters newswire topic classification task.

Dataset of 11,228 newswires from Reuters, labeled over 46 topics. 

As with the IMDB dataset, each wire is encoded as a sequence of word indexes (same conventions).

In [1]:
reset -fs

In [2]:
import numpy as np
import keras

Using TensorFlow backend.


In [3]:
from keras.datasets import reuters

In [4]:
print('Loading data...')
max_words = 1000
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=max_words,
                                                         test_split=0.2)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

Loading data...
8982 train sequences
2246 test sequences


In [5]:
from keras.preprocessing.text import Tokenizer

In [6]:
print('Vectorizing sequence data...')
max_words = 1000
tokenizer = Tokenizer(num_words=max_words)
x_train = tokenizer.sequences_to_matrix(x_train, mode='binary')
x_test = tokenizer.sequences_to_matrix(x_test, mode='binary')
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

Vectorizing sequence data...
x_train shape: (8982, 1000)
x_test shape: (2246, 1000)


In [7]:
num_classes = np.max(y_train) + 1
print(num_classes, 'classes')

46 classes


In [8]:
print('Convert class vector to binary class matrix '
      '(for use with categorical_crossentropy)')
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)
print('y_train shape:', y_train.shape)
print('y_test shape:', y_test.shape)

Convert class vector to binary class matrix (for use with categorical_crossentropy)
y_train shape: (8982, 46)
y_test shape: (2246, 46)


In [9]:
from keras.models import Sequential
from keras.layers import Dense, Activation

In [11]:
print('Building model...')

model = Sequential()

model.add(Dense(units=100, activation='relu',input_shape=(1000,))) # TODO: Add Dense layer. How big should input_shape being?
model.add(Dense(units=100, activation='relu')) # TODO: Add Activation
model.add(Dense(units=46, activation='softmax')) # TODO: Add Dense layer that matches output shape?
#model.add(None) # TODO: Add Activation to "squashes" the activation values
model.summary()

Building model...
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_4 (Dense)              (None, 100)               100100    
_________________________________________________________________
dense_5 (Dense)              (None, 100)               10100     
_________________________________________________________________
dense_6 (Dense)              (None, 46)                4646      
Total params: 114,846.0
Trainable params: 114,846
Non-trainable params: 0.0
_________________________________________________________________


In [12]:
print("Training model...")

model.compile(loss='categorical_crossentropy', # TODO: What should the loss function be?
              optimizer='adam',
              metrics=['accuracy'])

batch_size = 32
epochs = 5

history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    verbose=True,
                    validation_split=0.1)

Training model...
Train on 8083 samples, validate on 899 samples
Epoch 1/5
8083/8083 [==============================] - 2s - loss: 1.5417 - acc: 0.6456 - val_loss: 1.1513 - val_acc: 0.7419

In [13]:
score, accuracy = model.evaluate(x_test, y_test,
                                batch_size=batch_size, 
                                verbose=True)

print('Test score: {:.3}'.format(score))
print('Test accuracy: {:.3}'.format(accuracy))

1728/2246 [======================>.......] - ETA: 0sTest score: 1.03
Test accuracy: 0.779


__TODO__: What is baseline peformance? How does this simple model do?

**Solution**:Baseline was 74% without training. After few epochs it reached to 78%. Model did well given we didn't have many to construct much deep network

__TODO__: Spend a little bit of time tuning hyperparameters. What do you find?

**Solution**: Adding "adam" optimizer and another dense layer improved accuracy

<br>
<br> 
<br>

----